File: codetandem/tandem.py
"""Core tandem coding loop functionality."""

import json
from pathlib import Path
from typing import Dict, Optional, Union

from codetandem.curriculum import Module
from codetandem.modules import get_module_by_id, load_modules
from codetandem.providers.base import BaseAIProvider
from codetandem.state import get_current_module_id, load_state


class ProjectContext:
    """Container for project state and module information."""

    def __init__(
        self, state: Dict, modules: list, current_module: Module, project_path: Path
    ):
        """
        Initialize project context.

        Args:
            state: Project state dictionary
            modules: List of all modules
            current_module: The current active module
            project_path: Path to the project root
        """
        self.state = state
        self.modules = modules
        self.current_module = current_module
        self.project_path = project_path

    @property
    def current_module_id(self) -> str:
        """Get current module ID."""
        return get_current_module_id(self.state)

    @property
    def current_objectives(self) -> list:
        """Get objectives for current module."""
        return self.current_module.objectives

    @property
    def project_tree(self) -> Dict:
        """Get project file tree."""
        return self.state.get("project_tree", {})


def load_project_context(
    project_path: Union[str, Path],
    modules_file: str = "modules.json",
    state_file: str = "codetandem.state.json",
) -> ProjectContext:
    """
    Load project context from modules.json and state.json.

    Args:
        project_path: Path to the project root directory
        modules_file: Name of modules file (default: modules.json)
        state_file: Name of state file (default: codetandem.state.json)

    Returns:
        ProjectContext object with all loaded information

    Raises:
        FileNotFoundError: If required files are not found
        ValueError: If files are invalid or current module not found
    """
    project_path = Path(project_path).resolve()

    # Check if project path exists
    if not project_path.exists():
        raise FileNotFoundError(f"Project path not found: {project_path}")

    # Load modules
    modules_path = project_path / modules_file
    if not modules_path.exists():
        raise FileNotFoundError(
            f"modules.json not found at {modules_path}. "
            f"Run 'codetandem init' first to initialize the project."
        )

    modules = load_modules(modules_path)

    if not modules:
        raise ValueError("No modules found in modules.json")

    # Load state
    state_path = project_path / state_file
    if not state_path.exists():
        raise FileNotFoundError(
            f"State file not found at {state_path}. "
            f"Run 'codetandem init' first to initialize the project."
        )

    state = load_state(state_path)

    # Get current module
    current_module_id = get_current_module_id(state)
    try:
        current_module = get_module_by_id(modules, current_module_id)
    except ValueError:
        raise ValueError(
            f"Current module '{current_module_id}' not found in modules.json. "
            f"The state file may be out of sync."
        )

    return ProjectContext(
        state=state,
        modules=modules,
        current_module=current_module,
        project_path=project_path,
    )


def get_next_objective_index(context: ProjectContext) -> Optional[int]:
    """
    Get the index of the next objective to work on.

    Args:
        context: Project context

    Returns:
        Index of next objective, or None if all completed
    """
    # For now, return the first objective
    # In future, this could track which objectives are completed
    if context.current_objectives:
        return 0
    return None


def format_module_info(context: ProjectContext) -> str:
    """
    Format module information for display.

    Args:
        context: Project context

    Returns:
        Formatted string with module info
    """
    lines = []
    lines.append(f"Module: {context.current_module.title}")
    lines.append(f"ID: {context.current_module_id}")

    if context.current_objectives:
        lines.append("\nObjectives:")
        for i, obj in enumerate(context.current_objectives, 1):
            lines.append(f"  {i}. {obj}")

    return "\n".join(lines)


def build_coding_prompt(
    context: ProjectContext,
    target_file: Optional[Path] = None,
    file_content: Optional[str] = None,
    objective_index: int = 0,
    skill_level: float = 0.0,
) -> str:
    """
    Build a comprehensive prompt for AI code generation.

    Args:
        context: Project context
        target_file: Optional target file path (relative to project)
        file_content: Optional current content of target file
        objective_index: Index of the objective to work on (default: 0)
        skill_level: User's skill score for current module (0-10)

    Returns:
        Formatted prompt string for the AI
    """
    objective = None
    if objective_index < len(context.current_objectives):
        objective = context.current_objectives[objective_index]

    prompt_parts = []

    # System context
    prompt_parts.append(
        "You are CodeTandem, an AI pair programming assistant that helps users learn by writing foundational code and leaving TODO tasks for them to complete."
    )
    prompt_parts.append("")

    # Current learning module
    prompt_parts.append(f"# Current Learning Module")
    prompt_parts.append(f"Module: {context.current_module.title}")
    prompt_parts.append(f"Module ID: {context.current_module_id}")
    prompt_parts.append("")

    # Learning objective
    if objective:
        prompt_parts.append(f"# Current Learning Objective")
        prompt_parts.append(f"{objective}")
        prompt_parts.append("")

    # All objectives for context
    if context.current_objectives:
        prompt_parts.append(f"# All Module Objectives")
        for i, obj in enumerate(context.current_objectives, 1):
            marker = "→" if i - 1 == objective_index else " "
            prompt_parts.append(f"{marker} {i}. {obj}")
        prompt_parts.append("")

    # Target file information
    if target_file:
        prompt_parts.append(f"# Target File")
        prompt_parts.append(f"File: {target_file}")
        prompt_parts.append("")

        if file_content is not None:
            prompt_parts.append(f"# Current File Content")
            prompt_parts.append("```")
            prompt_parts.append(file_content if file_content else "# Empty file")
            prompt_parts.append("```")
            prompt_parts.append("")

    # Task instructions with dynamic scaffolding based on skill level
    prompt_parts.append("# Your Task")
    prompt_parts.append("Write foundational code that:")
    prompt_parts.append(
        "1. Implements the basic structure needed for the learning objective"
    )
    prompt_parts.append("2. Includes a clear TODO comment for the user to complete")
    prompt_parts.append(
        "3. Leaves the most educationally valuable part for the user to implement"
    )
    prompt_parts.append("")

    # Dynamic scaffolding based on skill level
    prompt_parts.append("# Scaffolding Level")
    if skill_level < 3.0:
        # Beginner: High scaffolding with detailed help
        prompt_parts.append("User skill level: BEGINNER")
        prompt_parts.append("Provide DETAILED scaffolding:")
        prompt_parts.append("- Write comprehensive foundational code")
        prompt_parts.append("- Include detailed TODO comments with step-by-step hints")
        prompt_parts.append("- Provide code snippets or examples in the TODO comment")
        prompt_parts.append("- Break down complex tasks into smaller, manageable steps")
        prompt_parts.append("- Add inline comments explaining key concepts")
    elif skill_level < 7.0:
        # Intermediate: Medium scaffolding with goal-oriented guidance
        prompt_parts.append("User skill level: INTERMEDIATE")
        prompt_parts.append("Provide GOAL-ORIENTED scaffolding:")
        prompt_parts.append("- Write solid foundational code structure")
        prompt_parts.append("- Include clear TODO comments describing the goal")
        prompt_parts.append(
            "- Provide hints about the approach, but not full implementations"
        )
        prompt_parts.append("- Let the user figure out implementation details")
        prompt_parts.append("- Add comments only for non-obvious design decisions")
    else:
        # Advanced: Low scaffolding with conceptual guidance
        prompt_parts.append("User skill level: ADVANCED")
        prompt_parts.append("Provide CONCEPTUAL scaffolding:")
        prompt_parts.append("- Write minimal foundational code (interfaces, types)")
        prompt_parts.append("- Include high-level TODO comments describing the concept")
        prompt_parts.append("- Focus on architecture and design patterns")
        prompt_parts.append("- Let the user implement most of the functionality")
        prompt_parts.append("- Assume the user understands implementation details")
    prompt_parts.append("")
    prompt_parts.append("Format your response as JSON with this structure:")
    prompt_parts.append("{")
    prompt_parts.append('  "file_path": "relative/path/to/file.ext",')
    prompt_parts.append('  "code": "complete file content with TODO",')
    prompt_parts.append('  "todo_line": 10,')
    prompt_parts.append(
        '  "todo_task": "Brief description of what the user should implement",'
    )
    prompt_parts.append('  "explanation": "Brief explanation of what you implemented"')
    prompt_parts.append("}")
    prompt_parts.append("")
    prompt_parts.append("The TODO comment should be formatted as:")
    prompt_parts.append(f"// TODO: [{context.current_module_id}] <task description>")

    return "\n".join(prompt_parts)


class CodeModification:
    """Represents a code modification from the AI."""

    def __init__(
        self,
        file_path: str,
        code: str,
        todo_line: int,
        todo_task: str,
        explanation: str,
    ):
        """
        Initialize code modification.

        Args:
            file_path: Relative path to the file
            code: Complete file content
            todo_line: Line number where TODO is located
            todo_task: Description of the TODO task
            explanation: Explanation of what was implemented
        """
        self.file_path = file_path
        self.code = code
        self.todo_line = todo_line
        self.todo_task = todo_task
        self.explanation = explanation


def generate_code_with_ai(
    provider: BaseAIProvider,
    context: ProjectContext,
    target_file: Optional[Path] = None,
    file_content: Optional[str] = None,
    objective_index: int = 0,
    skill_level: float = 0.0,
) -> CodeModification:
    """
    Generate code using AI provider.

    Args:
        provider: AI provider instance
        context: Project context
        target_file: Optional target file path
        file_content: Optional current file content
        objective_index: Index of objective to work on
        skill_level: User's skill score for current module (0-10)

    Returns:
        CodeModification object

    Raises:
        ValueError: If AI response is invalid
    """
    # Build prompt with skill level
    prompt = build_coding_prompt(
        context, target_file, file_content, objective_index, skill_level
    )

    # Get AI response
    response = provider.generate_code_suggestion(prompt, temperature=0.7)

    # Parse response
    return parse_code_modification(response.content)


def parse_code_modification(response_text: str) -> CodeModification:
    """
    Parse AI response to extract code modification.

    Args:
        response_text: Raw text response from AI

    Returns:
        CodeModification object

    Raises:
        ValueError: If response format is invalid
    """
    # Try to extract JSON from response
    # AI might wrap it in markdown code blocks
    text = response_text.strip()

    # Remove markdown code blocks if present
    if text.startswith("```json"):
        text = text[7:]
    elif text.startswith("```"):
        text = text[3:]

    if text.endswith("```"):
        text = text[:-3]

    text = text.strip()

    # Parse JSON
    try:
        data = json.loads(text)
    except json.JSONDecodeError as e:
        raise ValueError(f"Invalid JSON in AI response: {e}")

    # Validate required fields
    required_fields = ["file_path", "code", "todo_line", "todo_task", "explanation"]
    missing = [field for field in required_fields if field not in data]

    if missing:
        raise ValueError(
            f"Missing required fields in AI response: {', '.join(missing)}"
        )

    # Create modification object
    return CodeModification(
        file_path=data["file_path"],
        code=data["code"],
        todo_line=int(data["todo_line"]),
        todo_task=data["todo_task"],
        explanation=data["explanation"],
    )


def apply_code_modification(
    modification: CodeModification, project_path: Path, dry_run: bool = False
) -> Path:
    """
    Apply code modification to the project.

    Args:
        modification: CodeModification object
        project_path: Path to project root
        dry_run: If True, don't actually write files

    Returns:
        Absolute path to the modified file

    Raises:
        ValueError: If file path is invalid
    """
    # Resolve file path
    file_path = project_path / modification.file_path

    # Ensure parent directory exists
    if not dry_run:
        file_path.parent.mkdir(parents=True, exist_ok=True)

    # Write the file
    if not dry_run:
        file_path.write_text(modification.code, encoding="utf-8")

    return file_path

File: src/utils/tandem.ts
/**
 * Core tandem coding loop functionality.
 *
 * Provides the main logic for AI-assisted pair programming with dynamic scaffolding.
 */

import { writeFile, mkdir } from 'fs/promises';
import { dirname, resolve } from 'path';
import type { Module } from './curriculum.js';
import { getModuleById, loadModules } from './modules.js';
import { BaseAIProvider } from '../providers/base.js';
import type { StateData } from '../types/state.js';
import { getCurrentModuleId, loadState } from './state.js';

export interface ProjectContext {
  state: StateData;
  modules: Module[];
  currentModule: Module;
  projectPath: string;
}

/**
 * Load project context from modules.json and state.json.
 *
 * @param projectPath Path to the project root directory
 * @param modulesFile Name of modules file (default: modules.json)
 * @param stateFile Name of state file (default: codetandem.state.json)
 * @returns ProjectContext object with all loaded information
 * @throws Error if required files are not found or invalid
 */
export async function loadProjectContext(
  projectPath: string,
  modulesFile: string = 'modules.json',
  stateFile: string = 'codetandem.state.json'
): Promise<ProjectContext> {
  const resolvedPath = resolve(projectPath);

  // Load modules
  const modulesPath = resolve(resolvedPath, modulesFile);
  const modules = await loadModules(modulesPath);

  if (modules.length === 0) {
    throw new Error('No modules found in modules.json');
  }

  // Load state
  const statePath = resolve(resolvedPath, stateFile);
  const state = await loadState(statePath);

  // Get current module
  const currentModuleId = getCurrentModuleId(state);
  const currentModule = getModuleById(modules, currentModuleId);

  return {
    state,
    modules,
    currentModule,
    projectPath: resolvedPath,
  };
}

/**
 * Get the index of the next objective to work on.
 *
 * @param context Project context
 * @returns Index of next objective, or null if all completed
 */
export function getNextObjectiveIndex(context: ProjectContext): number | null {
  // For now, return the first objective
  // In future, this could track which objectives are completed
  if (context.currentModule.objectives.length > 0) {
    return 0;
  }
  return null;
}

/**
 * Format module information for display.
 *
 * @param context Project context
 * @returns Formatted string with module info
 */
export function formatModuleInfo(context: ProjectContext): string {
  const lines: string[] = [];
  lines.push(`Module: ${context.currentModule.title}`);
  lines.push(`ID: ${context.currentModule.id}`);

  if (context.currentModule.objectives.length > 0) {
    lines.push('\nObjectives:');
    for (let i = 0; i < context.currentModule.objectives.length; i++) {
      lines.push(`  ${i + 1}. ${context.currentModule.objectives[i]}`);
    }
  }

  return lines.join('\n');
}

/**
 * Build a comprehensive prompt for AI code generation with dynamic scaffolding.
 *
 * @param context Project context
 * @param targetFile Optional target file path (relative to project)
 * @param fileContent Optional current content of target file
 * @param objectiveIndex Index of the objective to work on (default: 0)
 * @param skillLevel User's skill score for current module (0-10)
 * @param codingBias Scaffolding level (guided/balanced/independent)
 * @param featureDescription Optional feature description from user
 * @returns Formatted prompt string for the AI
 */
export function buildCodingPrompt(
  context: ProjectContext,
  targetFile?: string,
  fileContent?: string,
  objectiveIndex: number = 0,
  skillLevel: number = 0.0,
  codingBias: 'guided' | 'balanced' | 'independent' = 'balanced',
  featureDescription?: string
): string {
  const objective =
    objectiveIndex < context.currentModule.objectives.length
      ? context.currentModule.objectives[objectiveIndex]
      : null;

  const promptParts: string[] = [];

  // System context
  promptParts.push(
    'You are CodeTandem, an AI pair programming assistant that helps users learn by writing foundational code and leaving TODO tasks for them to complete.'
  );
  promptParts.push('');

  // Current learning module
  promptParts.push('# Current Learning Module');
  promptParts.push(`Module: ${context.currentModule.title}`);
  promptParts.push(`Module ID: ${context.currentModule.id}`);
  promptParts.push('');

  // Learning objective
  if (objective) {
    promptParts.push('# Current Learning Objective');
    promptParts.push(objective);
    promptParts.push('');
  }

  // All objectives for context
  if (context.currentModule.objectives.length > 0) {
    promptParts.push('# All Module Objectives');
    for (let i = 0; i < context.currentModule.objectives.length; i++) {
      const marker = i === objectiveIndex ? '→' : ' ';
      promptParts.push(`${marker} ${i + 1}. ${context.currentModule.objectives[i]}`);
    }
    promptParts.push('');
  }

  // Target file information
  if (targetFile) {
    promptParts.push('# Target File');
    promptParts.push(`File: ${targetFile}`);
    promptParts.push('');

    if (fileContent !== undefined) {
      promptParts.push('# Current File Content');
      promptParts.push('```');
      promptParts.push(fileContent || '# Empty file');
      promptParts.push('```');
      promptParts.push('');
    }
  }

  // Feature description if provided
  if (featureDescription) {
    promptParts.push('# Feature Request');
    promptParts.push(`The user wants to build: ${featureDescription}`);
    promptParts.push('');
  }

  // Task instructions with dynamic scaffolding based on skill level
  promptParts.push('# Your Task');
  if (featureDescription) {
    promptParts.push('Write foundational code that:');
    promptParts.push('1. Implements this feature in a way that aligns with the current learning objectives');
    promptParts.push('2. Balances AI-written code with user TODOs based on the skill level and coding bias');
    promptParts.push('3. Links TODOs to the current module objectives where appropriate');
    promptParts.push('4. Provides educational value by leaving the most relevant parts for the user to complete');
  } else {
    promptParts.push('Write foundational code that:');
    promptParts.push('1. Implements the basic structure needed for the learning objective');
    promptParts.push('2. Includes a clear TODO comment for the user to complete');
    promptParts.push('3. Leaves the most educationally valuable part for the user to implement');
  }
  promptParts.push('');

  // Dynamic scaffolding based on skill level and coding bias
  promptParts.push('# Scaffolding Level');
  promptParts.push(`User skill level: ${skillLevel.toFixed(1)}/10.0`);
  promptParts.push(`User coding bias: ${codingBias}`);

  if (codingBias === 'guided') {
    promptParts.push('Provide DETAILED scaffolding:');
    promptParts.push('- Write comprehensive foundational code');
    promptParts.push('- Include detailed TODO comments with step-by-step hints');
    promptParts.push('- Provide code snippets or examples in the TODO comment');
  } else if (codingBias === 'balanced') {
    promptParts.push('Provide BALANCED scaffolding:');
    promptParts.push('- Write solid foundational code structure');
    promptParts.push('- Include clear TODO comments describing the goal');
    promptParts.push('- Let the user figure out implementation details');
  } else {
    // independent
    promptParts.push('Provide CONCEPTUAL scaffolding:');
    promptParts.push('- Write minimal foundational code (interfaces, types)');
    promptParts.push('- Include high-level TODO comments describing the concept');
    promptParts.push('- Let the user implement most of the functionality');
  }
  promptParts.push('');

  promptParts.push('Format your response as JSON with this structure:');
  promptParts.push('{');
  promptParts.push('  "file_path": "relative/path/to/file.ext",');
  promptParts.push('  "code": "complete file content with TODOs and success criteria",');
  promptParts.push('  "todos": [');
  promptParts.push('    {');
  promptParts.push('      "id": "obj-1",');
  promptParts.push('      "objectiveIndex": 0,');
  promptParts.push('      "task": "Brief description of what the user should implement",');
  promptParts.push('      "successCriteria": [');
  promptParts.push('        "Specific criterion 1 that must be met",');
  promptParts.push('        "Specific criterion 2 that must be met"');
  promptParts.push('      ],');
  promptParts.push('      "line": 10');
  promptParts.push('    }');
  promptParts.push('  ],');
  promptParts.push('  "explanation": "Brief explanation of what you implemented"');
  promptParts.push('}');
  promptParts.push('');
  promptParts.push('**IMPORTANT TODO Format Rules:**');
  promptParts.push('');
  promptParts.push('1. Each TODO must use format: // TODO: [obj-X] <task description>');
  promptParts.push('   - Where X is the objective number (1-based index)');
  promptParts.push(`   - Example: // TODO: [obj-1] Implement password validation`);
  promptParts.push('');
  promptParts.push('2. Include success criteria as comments ABOVE each TODO:');
  promptParts.push('   ```');
  promptParts.push('   // SUCCESS CRITERIA for [obj-1]:');
  promptParts.push('   // - Must validate password is at least 8 characters');
  promptParts.push('   // - Must check for at least one uppercase letter');
  promptParts.push('   // - Must return boolean true/false');
  promptParts.push('   // TODO: [obj-1] Implement password validation');
  promptParts.push('   ```');
  promptParts.push('');
  promptParts.push('3. Link each TODO to a specific learning objective from the module');
  promptParts.push('');
  promptParts.push(
    '4. Success criteria should be measurable, specific, and tied to the learning objective'
  );

  return promptParts.join('\n');
}

export interface TodoMetadata {
  id: string; // e.g., "obj-1"
  objectiveIndex: number;
  task: string;
  successCriteria: string[];
  line: number;
}

export interface CodeModification {
  filePath: string;
  code: string;
  todos: TodoMetadata[];
  explanation: string;
  // Legacy fields for backward compatibility
  todoLine?: number;
  todoTask?: string;
}

/**
 * Parse AI response to extract code modification.
 *
 * @param responseText Raw text response from AI
 * @returns CodeModification object
 * @throws Error if response format is invalid
 */
export function parseCodeModification(responseText: string): CodeModification {
  let text = responseText.trim();

  // Remove markdown code blocks if present
  if (text.startsWith('```json')) {
    text = text.substring(7);
  } else if (text.startsWith('```')) {
    text = text.substring(3);
  }

  if (text.endsWith('```')) {
    text = text.substring(0, text.length - 3);
  }

  text = text.trim();

  // Parse JSON
  let data: Record<string, unknown>;
  try {
    data = JSON.parse(text) as Record<string, unknown>;
  } catch (e) {
    const error = e instanceof Error ? e.message : String(e);
    throw new Error(`Invalid JSON in AI response: ${error}`);
  }

  // Check for new format (with todos array)
  if ('todos' in data && Array.isArray(data.todos)) {
    // New format
    const requiredFields = ['file_path', 'code', 'todos', 'explanation'];
    const missing = requiredFields.filter((field) => !(field in data));

    if (missing.length > 0) {
      throw new Error(`Missing required fields in AI response: ${missing.join(', ')}`);
    }

    const todos = (data.todos as unknown[]).map((todo: any) => ({
      id: String(todo.id),
      objectiveIndex: Number(todo.objectiveIndex),
      task: String(todo.task),
      successCriteria: Array.isArray(todo.successCriteria)
        ? (todo.successCriteria as string[])
        : [],
      line: Number(todo.line),
    }));

    return {
      filePath: String(data.file_path),
      code: String(data.code),
      todos,
      explanation: String(data.explanation),
      // Legacy fields for backward compatibility
      todoLine: todos[0]?.line,
      todoTask: todos[0]?.task,
    };
  } else {
    // Legacy format (backward compatibility)
    const requiredFields = ['file_path', 'code', 'todo_line', 'todo_task', 'explanation'];
    const missing = requiredFields.filter((field) => !(field in data));

    if (missing.length > 0) {
      throw new Error(`Missing required fields in AI response: ${missing.join(', ')}`);
    }

    return {
      filePath: String(data.file_path),
      code: String(data.code),
      todos: [
        {
          id: 'obj-1', // Default ID for legacy format
          objectiveIndex: 0,
          task: String(data.todo_task),
          successCriteria: [],
          line: Number(data.todo_line),
        },
      ],
      explanation: String(data.explanation),
      todoLine: Number(data.todo_line),
      todoTask: String(data.todo_task),
    };
  }
}

/**
 * Generate code using AI provider.
 *
 * @param provider AI provider instance
 * @param context Project context
 * @param targetFile Optional target file path
 * @param fileContent Optional current file content
 * @param objectiveIndex Index of objective to work on
 * @param skillLevel User's skill score for current module (0-10)
 * @param codingBias Scaffolding level (guided/balanced/independent)
 * @param featureDescription Optional feature description from user
 * @returns CodeModification object
 * @throws Error if AI response is invalid
 */
export async function generateCodeWithAI(
  provider: BaseAIProvider,
  context: ProjectContext,
  targetFile?: string,
  fileContent?: string,
  objectiveIndex: number = 0,
  skillLevel: number = 0.0,
  codingBias: 'guided' | 'balanced' | 'independent' = 'balanced',
  featureDescription?: string
): Promise<CodeModification> {
  // Build prompt with skill level, coding bias, and feature description
  const prompt = buildCodingPrompt(
    context,
    targetFile,
    fileContent,
    objectiveIndex,
    skillLevel,
    codingBias,
    featureDescription
  );

  // Get AI response
  const response = await provider.generateCodeSuggestion(prompt, {
    temperature: 0.7,
  });

  // Parse response
  return parseCodeModification(response.content);
}

/**
 * Apply code modification to the project.
 *
 * @param modification CodeModification object
 * @param projectPath Path to project root
 * @param dryRun If true, don't actually write files
 * @returns Absolute path to the modified file
 */
export async function applyCodeModification(
  modification: CodeModification,
  projectPath: string,
  dryRun: boolean = false
): Promise<string> {
  // Resolve file path
  const filePath = resolve(projectPath, modification.filePath);

  if (!dryRun) {
    // Ensure parent directory exists
    await mkdir(dirname(filePath), { recursive: true });

    // Write the file
    await writeFile(filePath, modification.code, 'utf-8');
  }

  return filePath;
}
